### PR TITLE
DEV-1555 Print failure logs at the end of pipeline

### PR DESCRIPTION
**Pull request by pauldwolfe**
Adds a field to the stage outputs to cache a reference to the log blob in case of failure.

Little awkward as we need to explicitly add the log everywhere, but can't find another solution
with the current state of the StageOutputs. The lanes in the BwaAligner also make it a bit
tricky to make a really general solution.

### CODE REVIEW

File: cluster/src/main/java/com/hartwig/pipeline/report/FailureSummary.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.Storage;
+import com.hartwig.pipeline.PipelineState;
+import com.hartwig.pipeline.StageOutput;
+import com.hartwig.pipeline.execution.PipelineStatus;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FailureSummary {
```
Comment (nedleitch) [inline cluster/src/main/java/com/hartwig/pipeline/report/FailureSummary.java:15]: I can see the reasoning (it's a summary showing any failures) but it's not necessarily a failure. Could we call it `StageSummary` or `ExecutionSummary` or something similar?

File: cluster/src/main/java/com/hartwig/pipeline/execution/vm/BashStartupScript.java [hunk 2]
```diff
@@ -50,7 +50,7 @@ String asUnixString(StorageStrategy storageStrategy) {
                 "function die() {",
                 "  exit_code=$?",
                 "  echo \"Unknown failure: called command returned $exit_code\"",
-                format("  gsutil -m cp %s gs://%s", localLogFile, runtimeBucketName),
+                format("  gsutil -m cp %s gs://%s/results/", localLogFile, runtimeBucketName),
```
Comment (nedleitch) [inline cluster/src/main/java/com/hartwig/pipeline/execution/vm/BashStartupScript.java:53]: I thought we had a requirement that it live in the top-level?

File: cluster/src/main/java/com/hartwig/pipeline/resource/OverriddenReferenceGenome.java [hunk 3]
```diff
@@ -53,7 +53,7 @@ public String gridssBreakendPon() {
 
     @Override
     public String gridssBreakpointPon() {
-        return decorated.gridssBreakpointPon();
+        return decorated.gridssBreakendPon();
```
Comment (nedleitch) [inline cluster/src/main/java/com/hartwig/pipeline/resource/OverriddenReferenceGenome.java:56]: How is this related?

**Review by nedleitch — CHANGES_REQUESTED** [review]

File: cluster/src/main/java/com/hartwig/pipeline/report/FailureSummary.java (hunk 1 above)
Comment (pauldwolfe) [reply cluster/src/main/java/com/hartwig/pipeline/report/FailureSummary.java:15]: See your point, going with VmExecutionLogSummary

File: cluster/src/main/java/com/hartwig/pipeline/resource/OverriddenReferenceGenome.java (hunk 3 above)
Comment (pauldwolfe) [reply cluster/src/main/java/com/hartwig/pipeline/resource/OverriddenReferenceGenome.java:56]: Not, was used to introduce a failure

File: cluster/src/main/java/com/hartwig/pipeline/execution/vm/BashStartupScript.java (hunk 2 above)
Comment (pauldwolfe) [reply cluster/src/main/java/com/hartwig/pipeline/execution/vm/BashStartupScript.java:53]: don't think so, when the run succeeds this is where the logs go.
